Http: HttpEntity stream cancellation

ataraxer opened the pull request:
Here's an attempt to solve issues #18010 and #18203.

First commit can be treated as a simpler version of the solution, which involves adding `eagerClose` flag to `splitWhen` in order to define how substream cancelation should be treated: by draining the remaining elements in order to prevent stalling (current behaviour) or cancelling the stage altogether.

Second one introduces custom stage instead of `splitWhen/prefixAndTail` pipeline, which would allow to add `eagerClose` flag to `outgoingConnection` without adding flag to `splitWhen`/`splitAt` if this change isn't plausible.

I used https://github.com/akka/akka/commit/ced5aa7ddcac1408e08ed159da9a04fecfad18ed as a reference, however there are some slight changes which will be marked in comments to particular lines.

ataraxer commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -149,6 +133,68 @@ private[http] object OutgoingConnectionBlueprint {
+      }
+
+      override def preStart(): Unit = pull(in)
+      setHandler(out, eagerTerminateOutput)
```
https://github.com/akka/akka/commit/ced5aa7ddcac1408e08ed159da9a04fecfad18ed#diff-623ef36775342f2db6b86855f7ffea64R122 uses custom `OutHandler` which just pulls the input, however while testing I've discovered that this leads to port being pulled twice:

```
[ERROR] [01/20/2016 00:58:08.384] [http-streaming-akka.actor.default-dispatcher-4] [ActorSystem(http-streaming)] Outgoing request stream error
java.lang.IllegalArgumentException: requirement failed: Cannot pull port twice
 at scala.Predef$.require(Predef.scala:233)
 at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:420)
 at akka.http.impl.engine.client.OutgoingConnectionBlueprint$ResponsePrep$$anon$3$$anon$5.onPull(OutgoingConnectionBlueprint.scala:185)
 ...
```

If I understand it correctly this happens because of handler which is set up on `SubSourceOutlet` for entity, which also produces pull's from the same stage.

drewhk replied:
that might be right, yes

drewhk commented:
OK TO TEST

ktoso commented:
OK TO TEST

drewhk commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -346,9 +349,12 @@ final class Split[T](decision: Split.SplitDecision, p: T ⇒ Boolean) extends Gr
-            // Drain into the void
-            if (substreamCancelled) pull(in)
-            else substreamSource.push(elem)
+            if (substreamCancelled) {
```
is it actually what you want? I though you want to cancel the upstream and complete possible substreams if the master stream is cancelled. Currently it cancels the upstream and everything when a substream is cancelled. But AFAIK substreams *are* the entitiy streams already, so if someone gets to cancel them that means they are properly consumed.

ataraxer replied:
Yes, this is precisely the case that was discussed in both https://github.com/akka/akka/issues/18010:

> To be clear, the desired behavior is that cancelling the response entity should result in a RST being sent to the server and the entire TCP connection being torn down.

and https://github.com/akka/akka/issues/18203:

> In general, I would expect Sink.cancelled to at least free the connection. With respect to your question whether it should abort the connection or just silently consume the entity, I think that probably it would be nice when we could save sending the response over TCP at all, but I have no idea what that means in terms of HTTP (maybe an error on the server side?). Silently consuming the entity on the other hand would also be unexpected when I cancel something. So I would argue that it should close/abort the connection, without knowing the implications for things like keep-alive or pipelining.

To clarify things a bit: I'm not working in the same company as @mspielberg, but we have a very similar use case and cancelling TCP is precisely what we need when response entity is cancelled (e.g. by running with `Sink.cancelled`) or only consumer only partially (e.g. via `take(n)`).

In our case we have an infinite response stream for consuming data from message queue service. This request also acquires a lock on the server side (this has to do with at least once delivery guarantees), which is automatically released when connection is closed. I wrote a client for this service in akka-http which wraps underlying HTTP protocol and exposes data stream in variety of ways for different use cases, the simplest one being a simple `Source[Message]`. The user may choose to consume only a limited amount of messages from this source, which will translate into entity of an underlying chunked `HttpResponse` being cancelled after only required amount of messages has been consumed. The problem is that current implementation keeps draining chunks  in order to prevent connection stalling, since all requests are occupying the same connection, but since the stream is infinite we end up just reading data into /dev/null consuming excessive resources (not good) and also not releasing the lock on the stream and thus preventing anyone else from accessing it (very-very bad).

ataraxer commented:
I've updated the `ResponsePrep` stage -- now it should pass all tests but one:
```
ConnectionPoolSpec:
 The host-level client infrastructure
 - should open a second connection if the request on the first one is dispatched but not yet completed *** FAILED *** (3 seconds, 163 milliseconds)
```

For some reason chunk which is sent through the first response isn't arriving from TCP output stream. However if `responseEntitySub.request(1)` is removed it starts to appear in logs. I'm still getting used to streaming internals so I might need some help on this one. I will try to investigate further in a meantime.

rkuhn commented on akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala:
```diff
@@ -346,9 +349,12 @@ final class Split[T](decision: Split.SplitDecision, p: T ⇒ Boolean) extends Gr
-            if (substreamCancelled) pull(in)
-            else substreamSource.push(elem)
+            if (substreamCancelled) {
+              if (eagerClose) completeStage()
```
shouldn’t this rather happen in onDownstreamFinish for the SubSource?

ataraxer replied:
No, since we need to read at least one more message to find out whether substream has been fully consumed (if next message is matching predicate) or interrupted in the middle, since otherwise normal completion of first substream will always lead to master stream being cancelled on `eagerClose`.

rkuhn replied:
Hmm, not sure I completely understand: receiving cancellation from downstream should only happen if the downstream decides that it does not want more data, which means that canceling is the desired semantics in the eagerClose case (which BTW should be renamed to eagerCancel). Normal substream termination due to the next element matching the substream will signal `substreamSource.complete()`, inhibiting further reception of `onDownstreamFinish()`, so in that case all should be good.

The only “interesting” case is if the substream processing logic double-checks the length of the stream and terminates of its own accord just at the point where the stream would otherwise have ended—but I consider that a less important case than to pass along cancellation without needing further upstream data; an open TCP connection binds resources even when no data are flowing.

ataraxer replied:
The problem is that there is no distinction between mid-stream cancelation and completion of a stream by fully consuming it, both are resulting as `onDownstreamFinish` callback (unless I'm missing something?).

This means that if we will complete stage on `onDownstreamFinish` of a substream when `eagerClose` flag is set, the stream will always terminate as soon as first substream completes (disregarding whether it was interrupted or fully consumed).

Here's an example:

```scala
Source(0 to 9000)
  .splitWhen(0 % 100 == 0, eagerClose = true)
  .concatSubstreams
```

Here each substream will be fully consumed, resulting in `onDownstreamFinish` after the last element of a substream is consumed. So if we complete the entire stage in `onDownstreamFinish` the stream will be interrupted after the first 100 elements.

The purpose of `eagerClose` is to avoid situations when substreams are cancelled without fully draining them, which currently leads to elements being drained automatically which may be unwelcome in case of big or even infinite streams or cases when items of a stream are tightly coupled with some resource consuming work (e.g. HTTP). However, if the substream has been fully consumed there is no point in tearing down the stage.

drewhk replied:
It is `onUpstreamFinish` which is signalled when the last element of a substream is consumed. There is no `onDownstreamFinish` on a port which has been closed, which is usually what happens after `onUpstreamFinish`. The only situations where a stage will get a `onDownstreamFinish` if the consumer does not want to read more. See http://doc.akka.io/docs/akka-stream-and-http-experimental/2.0.2/scala/stream-customize.html#Port_states__InHandler_and_OutHandler

ataraxer replied:
> which BTW should be renamed to eagerCancel

I've seen both being used in different stages/combinators, is there a clear distinction between them? `eagerCancel` is for behaviour related to downstream completion and `eagerClose` is for upstream, correct?

> It is onUpstreamFinish which is signalled when the last element of a substream is consumed.

Ahh, I see.

rkuhn replied:
> I've seen both being used in different stages/combinators, is there a clear distinction between them? eagerCancel is for behaviour related to downstream completion and eagerClose is for upstream, correct?

We have renamed that parameter on the Merge operations to be more precise: there is it about emitting the completion signal, here it is about emitting the cancellation signal.

ataraxer replied:
OK. Should it also be renamed on `outgoingConnection`? I think `eagerClose` suits it better, since it will translate into connection being closed.

rkuhn replied:
WDYM? `Tcp.outgoingConnection` has a `halfClose` parameter, which means something different. “close” is not properly defined for a stage, we have _complete_, _fail_, _cancel_ signals.

We still have eagerClose in SslTlsStage (to be rewritten soon) and on Interleave (which should be changed to eagerComplete).

ataraxer replied:
In TCP/HTTP terminology connection is being "closed", so maybe for HTTP client flow constructors `eagerClose` would be more clear. On the other hand I see that `eagerCancel` would be more consistent with other stages.

rkuhn commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1093,16 +1093,18 @@ trait FlowOps[+Out, +Mat] {
    * See also [[FlowOps.splitAfter]].
    */
-  def splitWhen(p: Out ⇒ Boolean): SubFlow[Out, Mat, Repr, Closed] = {
+  def splitWhen(p: Out ⇒ Boolean, eagerClose: Boolean = false): SubFlow[Out, Mat, Repr, Closed] = {
```
this will be needed for splitAfter as well

ataraxer replied:
Sure, I just wanted to decide on API and whether those changes are OK or not first.

ataraxer replied:
What do you thing about the signature? Positional elements after lambda look kinda ugly, so maybe an overload like:
```scala
def splitWhen(eagerClose: Boolean)(p: Out => Boolean)
```
should be introduced instead? This also has the advantage of being extendible with other options/flags in the future, while allowing to use a block syntax for lambda:
```scala
Source(0 to 10).splitWhen(eagerClose = true) { item =>
  ...
  ...
  ...
}
```

rkuhn commented on akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala:
```diff
@@ -241,16 +243,17 @@ class HttpExt(private val config: Config)(implicit val system: ActorSystem) exte
    */
-  def clientLayer(hostHeader: Host): ClientLayer =
-    clientLayer(hostHeader, ClientConnectionSettings(system))
+  def clientLayer(hostHeader: Host, eagerClose: Boolean): ClientLayer =
```
why should this be configurable? HTTP is rather specific in what it needs, so we can just derive this

rkuhn replied:
ah, yes, I see now; for the client this is actually required

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -149,6 +133,90 @@ private[http] object OutgoingConnectionBlueprint {
+
+                case messagePart ⇒
+                  if (responseCancelled) {
+                    if (eagerClose) completeStage()
```
all completions must be accompanied by closing the entitySource as well, else we leak it

rkuhn replied:
This means that onUpstreamFailure also needs to be overridden.

rkuhn commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -70,24 +71,7 @@ private[http] object OutgoingConnectionBlueprint {
-          SubSource.kill(tail)
-          throw IllegalResponseException(info)
-      }.concatSubstreams
+      .via(new ResponsePrep(parserSettings, eagerClose))
```
Great performance improvement as well!

ktoso replied:
Awesome, thanks :-)

ataraxer commented:
What do you think about the signature? Positional elements after lambda look kinda ugly, so maybe an overload like:
```scala
def splitWhen(eagerClose: Boolean)(p: Out => Boolean)
```
should be introduced instead? This also has the advantage of being extendable with other options/flags in the future, while allowing to use a block syntax for lambda:
```scala
Source(0 to 10).splitWhen(eagerClose = true) { item =>
  ...
  ...
  ...
}
```

ataraxer commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -149,6 +133,98 @@ private[http] object OutgoingConnectionBlueprint {
+        creator match {
+          case StrictEntityCreator(entity) ⇒
+            pull(in)
+            setHandler(in, waitForMessageEnd)
```
Is there probably a better way to read exactly one message of expected type/value?

ktoso replied:
This style is fine IMO. We don't have a specific method for the scenario given (you may want to look at `read`, but it's not really better than just setting a handler like already done here).

ataraxer commented:
- renamed `eagerClose -> eagerCancel` for combinators;
- moved `eagerCancel` handling to `onDownstreamFinish`;
- added `eagerCancel` for `splitAfter`;
- added test cases for `eagerCancel`;
- added `onUpstreamFailure` handler for substreams of `ResponsePrep`;
- added better handling of unexpected items in `ResponsePrep` stage states;

Tests for TCP connection being closed on `HttpResponse` entities cancellation are still WIP (looking for a proper suit to put them into).

Still no luck in figuring out what is wrong with `ConnectionPoolSpec` test 😕

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -149,6 +133,98 @@ private[http] object OutgoingConnectionBlueprint {
+      private val waitForMessageEnd = new InHandler {
+        def onPush(): Unit = grab(in) match {
+          case MessageEnd ⇒ setHandler(in, idle)
+          case other      ⇒ throw new IllegalStateException(f"MessageEnd expected but $other received.")
```
We use the `s` interpolator usually, it's enough here too

rkuhn commented:
LGTM @ataraxer ! I agree with the reasoning concerning the order of the arguments in the Scala DSL, where currying provides nice readability; please add the overloads as proposed (reverting the current signatures to their original state) and then we can merge.

ConnectionPoolSpec is a different issue that we’ll need to figure out, it is not related to this PR AFAICS.

ataraxer commented:
Well, it used to work with original implementation just fine (: I think I might have actually figured it out now.

And what about the "`eagerClose`/`eagerCancel` on `outgoingConnection` controversy"? (: What's the final decision?

ataraxer commented:
OK, almost there.
##### TODO
- decide on a flag name;
- add flag to connection pools as well;
- tests, scaladoc and documentation for eagerly closed HTTP connections.

ataraxer committed:
+str #19528 Add `propagateSubstreamCancel` flag to `splitWhen` and `splitAfter`

ataraxer committed:
=htc #19528 OutgoingConnectionBlueprint: replace splitWhen/prefixAndTail with custom stage

ataraxer commented:
Also, should eager close flag be added to connection pool settings or only be available programmatically?

ataraxer commented:
ping

johanandren commented:
Hi @ataraxer I'm afraid we are spread very thin this week, so that is causing the lack of feedback. We'll look at this as soon as we can!

ataraxer commented:
Ping 2: The Pingening

How about this week? (:

drewhk commented:
First of all, it will need a rebase :)

ataraxer commented:
Ping 3: Revenge of the Ping

rkuhn commented:
LGTM

rkuhn commented:
@drewhk please merge tomorrow morning after a last look, I don’t trust my brain right now.

ataraxer commented:
Please note that there are still some open questions above.

drewhk commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -148,6 +132,104 @@ private[http] object OutgoingConnectionBlueprint {
+        }
+      }
+    }
+  }
```
This stage looks correct. Nice work!

ataraxer replied:
Please give it another quick look, I moved things around a bit to avoid unnecessary allocations.

ktoso replied:
I gave it another look and looks good

drewhk commented:
My opinions:
 - first of all, great work!
 - as for the flag, I would call it `propagateSubstreamCancel` for `splitWhen` 
 - as for HTTP, I think we might not need this to be configurable at all! If someone wants to retain the original behavior, he can just attach a Sink.ignore which will not cancel the substream but drain it into the nothingness. Sink.cancel will automatically close the connection, and I gues that is what people expect. In general, the previous behavior (not propagate cancel) was impossible to opt out from, while with reversing the default it is possible to opt out even without a flag, just by using the proper Sink.

drewhk commented:
Also, if we agree on the default propagate-cancel, we might want to implement it for the server as well.

ataraxer commented:
Thanks!

Hopefully I would be able to clean things up this evening and take a look at server side as well.

drewhk commented:
Also, the benefit of the approach I sketched is that compared to a static flag, the appropriate action can be determined case-by-case, simply by using the proper Sink according to some criteria.

ataraxer committed:
!htp #19528 HttpServerBluePrint: close connection on request entity cancellation

ataraxer commented:
Aaaaand done.

drewhk commented:
Almost, the migration guide needs to mention:
 - the actual change, what is the behavior now
 - how to obtain the original behavior (a code snippet showing Sink.ignore - preferably pulled out from an actual unit test).

ataraxer commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -151,8 +151,16 @@ private[http] object HttpServerBluePrint {
 
         val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
-        // optimization: re-use the idle outHandler
-        entitySource.setHandler(this)
```
This optimization can be restored in a similar way to client's `ResponsePrep` stage by moving entity stream to a variable and implementing `onDownstreamFinish` as a member of stage itself. Things look a bit ..."convoluted" on the server side though, so I decided to avoid messing with it for now (:

ktoso replied:
Can you put a TODO with a note for this here?

ataraxer commented on akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala:
```diff
@@ -151,8 +151,16 @@ private[http] object HttpServerBluePrint {
+          }
+        })
 
         setHandler(in, new InHandler {
```
btw this handle can also be combined with the `OutHandler` which is being set on `out` port below too.

ataraxer commented:
Fixed documentation and added tests.

drewhk commented:
There are Java docs as well ;)

ataraxer commented:
Done.

drewhk commented:
LGTM! Great contribution!

ktoso commented:
I'm very interested in this one, will review today after some more catching up on other things. Looks very promising :)

drewhk commented:
Ok, let's try to merge this today or latest tomorrow so we get enough Jenkins runs on it until Friday.

ktoso commented on akka-stream/src/main/scala/akka/stream/scaladsl/Flow.scala:
```diff
@@ -1202,25 +1202,38 @@ trait FlowOps[+Out, +Mat] {
    * See also [[FlowOps.splitAfter]].
    */
-  def splitWhen(p: Out ⇒ Boolean): SubFlow[Out, Mat, Repr, Closed] = {
+  def splitWhen(propagateSubstreamCancel: Boolean)(p: Out ⇒ Boolean): SubFlow[Out, Mat, Repr, Closed] = {
```
Signature is nice this way :+1: Now we're missing a JavaDSL correspondent for it though. 
Please add a similar one to javadsl.Flow

ktoso replied:
Technically we started phasing out booleans with special classes to make the APIs more readable, so in javadsl it would follow that pattern instead of a boolean I think.

ktoso commented on akka-docs/rst/java/http/client-side/connection-level.rst:
```diff
@@ -53,6 +53,10 @@ The connection can also be closed by the server.
 An application can actively trigger the closing of the connection by completing the request stream. In this case the
 underlying TCP connection will be closed when the last pending response has been received.
 
+Connection will also be closed if response entity has been cancelled (e.g. by attaching it to ``Sink.cancelled()``)
```
"The connection will also be closed if the response entity is cancelled" IMHO sounds more right

ktoso commented on akka-docs/rst/scala/stream/migration-guide-2.0-2.4-scala.rst:
```diff
@@ -102,6 +102,19 @@ and were accessible via ``settings``. We now made it possible to configure the p
+Client / server behaviour on cancelled entity
+---------------------------------------------
+
+Previously if request received by server-side or response received by client-side were
```
if *a* request received

ktoso replied:
Maybe its enough to say: Previously if request or response were cancelled or consumer only partially...

ktoso commented on akka-docs/rst/scala/stream/migration-guide-2.0-2.4-scala.rst:
```diff
@@ -102,6 +102,19 @@ and were accessible via ``settings``. We now made it possible to configure the p
+more requests / responses incoming. Now the default behaviour is to close the connection
+in order to prevent using excessive resource usage in case of huge entities.
+
+Old behaviour can be achieved by explicitly draining the entity:
```
The old behaviour

ktoso commented on akka-http-core/src/main/scala/akka/http/impl/engine/client/OutgoingConnectionBlueprint.scala:
```diff
@@ -69,24 +69,7 @@ private[http] object OutgoingConnectionBlueprint {
     val responsePrep = Flow[List[ResponseOutput]]
       .mapConcat(conforms)
-      .splitWhen(x ⇒ x.isInstanceOf[MessageStart] || x == MessageEnd)
-      .prefixAndTail(1)
```
Great to get rid of this one :)

ktoso commented:
LGTM, had some minor spelling comments but code looks good, thanks a lot @ataraxer!

2beaucoup commented on akka-docs/rst/scala/http/client-side/connection-level.rst:
```diff
@@ -55,6 +55,9 @@ The connection can also be closed by the server.
 
+Connection will also be closed if response entity has been cancelled (e.g. by attaching it to ``Sink.cancelled``)
+or consumed only partially (e.g. by calling ``take()`` on it). In order to prevent this behaviour entity should be
+explicitly drained by attaching it to ``Sink.ignore``.
```
Maybe sprinkle some 'a's or 'an's to prefix nouns like 'connection' and 'entity' (2x)? Or just 'Connections'?

ktoso replied:
:+1: had a comment around the same thing https://github.com/akka/akka/pull/19528#discussion_r52298380

ataraxer commented:
Argh, missed one.

ataraxer commented:
Oooook, now everything should be in order:

- restored server optimizations;
- fixed docs;
- rebased.

ktoso commented:
Thanks a lot - great job on this one! Let's merge after validation :)

drewhk commented:
Great work!

ataraxer commented:
Thanks!

Will there be another 2.0.x release or this is going to be a part of 2.4.2?

drewhk commented:
Bugfixes go into 2.0.x but features and non-compatible changes do not. This is unfortunately a breaking change.

ataraxer commented:
OK, thanks for clarification.

johanandren commented:
PLS BUILD

ktoso commented:
Thanks again, great contrib! :)
